tidy3d/components/medium.py

```diff
@@ -548,208 +1007,142 @@ def eps_diagonal_on_grid(
         """
-        field_components = {field_name: eps.copy() for field_name in ("eps_xx", "eps_yy", "eps_zz")}
-        eps_dataset = PermittivityDataset(**field_components)
-        return cls(eps_dataset=eps_dataset, interp_method=interp_method)
+        if isinstance(eps, SpatialDataArray):
+            # purely real, not need to know `freq`
+            if CustomMedium._validate_isreal_dataarray(eps):
+                return cls(permittivity=eps, interp_method=interp_method)
+            # complex permittivity, needs to know `freq`
+            if freq is None:
+                raise SetupError(
+                    "For a complex 'eps', 'freq' at which 'eps' is defined must be supplied",
+                )
+            eps_inf, sigma = CustomMedium.eps_complex_to_eps_sigma(eps, freq)
+            return cls(permittivity=eps_inf, conductivity=sigma, interp_method=interp_method)
+
+        # eps is ScalarFieldDataArray
+        # contradictory definition of frequency
+        freq_data = eps.coords["f"].data[0]
+        if freq is not None and not isclose(freq, freq_data):
+            raise SetupError(
+                "'freq' value is inconsistent with the coordinate 'f'"
+                "in 'eps' DataArray. It's unclear at which frequency 'eps' "
+                "is defined. Please leave 'freq=None' to use the frequency "
+                "value in the DataArray."
+            )
+        eps_inf, sigma = CustomMedium.eps_complex_to_eps_sigma(eps, freq_data)
+        eps_inf = SpatialDataArray(eps_inf.squeeze(dim="f", drop=True))
+        sigma = SpatialDataArray(sigma.squeeze(dim="f", drop=True))
+        return cls(permittivity=eps_inf, conductivity=sigma, interp_method=interp_method)
 
     @classmethod
     def from_nk(
         cls,
-        n: ScalarFieldDataArray,
-        k: Optional[ScalarFieldDataArray] = None,
+        n: Union[ScalarFieldDataArray, SpatialDataArray],
+        k: Optional[Union[ScalarFieldDataArray, SpatialDataArray]] = None,
+        freq: float = None,
         interp_method: InterpMethod = "nearest",
     ) -> CustomMedium:
         """Construct a :class:`.CustomMedium` from datasets containing n and k values.
 
         Parameters
         ----------
-        n : :class:`.ScalarFieldDataArray`
+        n : Union[:class:`.SpatialDataArray`, :class:`.ScalarFieldDataArray`]
             Real part of refractive index.
-        k : :class:`.ScalarFieldDataArray` = None
+        k : Union[:class:`.SpatialDataArray`, :class:`.ScalarFieldDataArray`], optional
             Imaginary part of refrative index.
+        freq : float, optional
+            Frequency at which ``n`` and ``k`` are defined.
         interp_method : :class:`.InterpMethod`, optional
-                Interpolation method to obtain permittivity values that are not supplied
-                at the Yee grids.
+            Interpolation method to obtain permittivity values that are not supplied
+            at the Yee grids.
 
```

Comment:
I think we should add a `Note` explaining briefly how it works with `freq` and the types of `n` `k`. for example:
```
If both `n` and `k` are supplied without any frequency coordinate, the `freq` kwarg will be used to evaluate the permittivity and conductivity.
```

Reply:
How about

```
For lossy medium, if both ``n`` and ``k`` are supplied through :class:`.SpatialDataArray` that doesn't contain frequency information, the ``freq`` kwarg will be used to evaluate the permittivity and conductivity. Alternatively, ``n`` and ``k`` can be supplied through :class:`.ScalarFieldDataArray` with frequency coordinate, and leave ``freq`` kwarg as default ``None``.
```

Reply:
edited a bit but looks good overall.

```
For lossy medium, if both ``n`` and ``k`` are supplied through :class:`.SpatialDataArray`, which doesn't contain frequency information, the ``freq`` kwarg will be used to evaluate the permittivity and conductivity. Alternatively, ``n`` and ``k`` can be supplied through :class:`.ScalarFieldDataArray`, which contains a frequency coordinate. In this case, leave ``freq`` kwarg as the default of ``None``.
```